dhis-2/dhis-api/src/main/java/org/hisp/dhis/common/TranslatableObject.java

```diff
@@ -0,0 +1,142 @@
+import com.fasterxml.jackson.dataformat.xml.annotation.JacksonXmlProperty;
+
+/**
+ * Base class for translatable object.
+ * <p>
+ * Contains necessary methods for storing and retrieving translation values.
+ * <p>
+ * Beside extending this object, translatable object needs to have a column
+ * translations with type jblTranslations defined in hibernate mapping file.
+ */
+public class TranslatableObject
+{
+    /**
+     * Set of available object translation, normally filtered by locale.
+     */
+    protected Set<Translation> translations = new HashSet<>();
+
+    /**
+     * Cache for object translations, where the cache key is a combination of
+     * locale and translation property, and value is the translated value.
+     */
+    protected Map<String, String> translationCache = new HashMap<>();
+
+    // -------------------------------------------------------------------------
+    // Getters and setters
+    // -------------------------------------------------------------------------
+
+    @Gist( included = Gist.Include.FALSE )
+    @JsonProperty
+    @JacksonXmlElementWrapper( localName = "translations", namespace = DxfNamespaces.DXF_2_0 )
+    @JacksonXmlProperty( localName = "translation", namespace = DxfNamespaces.DXF_2_0 )
+    public Set<Translation> getTranslations()
+    {
+        if ( translations == null )
+        {
+            translations = new HashSet<>();
+        }
+
+        return translations;
+    }
+
+    /**
+     * Clears out cache when setting translations.
+     */
+    public void setTranslations( Set<Translation> translations )
+    {
+        this.translationCache.clear();
+        this.translations = translations;
+    }
+
+    // -------------------------------------------------------------------------
+    // Util methods
+    // -------------------------------------------------------------------------
+
+    /**
+     * Returns a translated value for this object for the given property. The
+     * current locale is read from the user context.
+     *
+     * @param translationKey the translation key.
+     * @param defaultValue the value to use if there are no translations.
+     * @return a translated value.
+     */
+    protected String getTranslation( String translationKey, String defaultValue )
+    {
+        Locale locale = UserContext.getUserSetting( UserSettingKey.DB_LOCALE );
+
+        defaultValue = defaultValue != null ? defaultValue.trim() : null;
+
+        if ( locale == null || translationKey == null )
+        {
+            return defaultValue;
+        }
+
+        loadTranslationsCacheIfEmpty();
+
+        String cacheKey = Translation.getCacheKey( locale.toString(), translationKey );
+
+        return translationCache.getOrDefault( cacheKey, defaultValue );
+    }
+
+    /**
+     * Populates the translationsCache map unless it is already populated.
+     */
+    private void loadTranslationsCacheIfEmpty()
```

Comment:
This needs some way to make it thread-safe. Either the method needs a thread-guard or the map is loading per key utilizing `translationCache.computeIfAbsent`. Another way would be to always eagerly update the cache each time the `setTranslations` is called.